jackknife additions and fixes, numba acceleration

HDembinski opened the pull request:
* compute jackknife with numba
* fixes jackknife_bias formula
* adds jackknife_bias_corrected
* possibly faster and more accurate jackknife_variance
* added unit tests for jackknife functions
* added docu

HDembinski commented on src/resample/tests/test_bootstrap.py:
```diff
@@ -17,9 +19,54 @@
+    # f = 5/3
+    # (n-1) * (mean_loo_f - f)
+    # (3 - 1) * (6/3 - 5/3) = 2/3
+    assert r == pytest.approx(2.0 / 3.0)
```
assert_almost_equal would also work. I slightly prefer pytest.approx for testing single numbers, since assert_equal and assert_almost_equal are primarily designed for arrays.

HDembinski replied:
pytest.approx is the intended pytest facility for testing floating point results.

HDembinski replied:
It is a matter of taste.

HDembinski commented on src/resample/tests/test_bootstrap.py:
```diff
@@ -17,9 +19,54 @@
+    assert r == 0
+
+    # this "mean" has a bias of exactly O(n^{-1})
+    def bad_mean(x):
```
For the record, I don't see a real advantage in fine-grained test functions and it feels more messy to me. But it is your project, so I will split this up.

dsaxton replied:
It's partly just preference, but one reason is that the test will stop execution after the first failed assertion, so some tests may not run if there are several in one test function. For example with something like
```python
def test_dummy():
    assert False
    # more failing tests
```
we'll only see the first failed assertion.

pep8speaks commented:
Hello @HDembinski! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`setup.py`](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/setup.py):

> [Line 9:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/setup.py#L9): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

* In the file [`src/resample/bootstrap.py`](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py):

> [Line 40:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L40): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 47:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L47): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 48:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L48): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 49:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L49): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 50:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L50): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 59:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L59): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 74:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L74): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 86:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L86): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 114:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L114): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 139:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L139): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)

HDembinski commented:
> Can you also merge master (might be why the CI didn't run, and you may need to add numba to requirements since the CI uses that file)?

I added numba to requirements, and changed setup.py so that it uses the info from requirements.txt. I cannot merge master since this branch branches of from the master head.

dsaxton commented:
> Hello @HDembinski! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:
> 
> * In the file [`setup.py`](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/setup.py):
> 
> > [Line 9:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/setup.py#L9): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> 
> * In the file [`src/resample/bootstrap.py`](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py):
> 
> > [Line 40:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L40): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> > [Line 47:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L47): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> > [Line 48:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L48): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> > [Line 49:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L49): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> > [Line 50:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L50): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> > [Line 59:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L59): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> > [Line 74:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L74): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> > [Line 86:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L86): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> > [Line 114:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L114): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> > [Line 139:80](https://github.com/dsaxton/resample/blob/9b389728d5c6a01708118b58986d31cd7079ff51/src/resample/bootstrap.py#L139): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)

@HDembinski You can ignore this, I got rid of it on master (I'm guessing it's triggering because the YAML file is on this branch)

HDembinski commented:
Can we please increase the maximum line width so something more modern, like 88 or 90 characters? Raymond Hettinger even recommends 120, which I find too much. 90 is a compromise.

HDembinski commented:
> @HDembinski You can ignore this, I got rid of it on master (I'm guessing it's triggering because the YAML file is on this branch)

Ah, ok :)

dsaxton commented on src/resample/bootstrap.py:
```diff
@@ -111,12 +111,12 @@ def jackknife_bias_corrected(a: np.ndarray, f: Callable) -> np.ndarray:
 
     f : callable
-        Estimator. Can be any mapping R^N -> R^M, where N is the number of samples.
+        Estimator. Can be any mapping ℝⁿ → ℝᵐ, where n is the number of samples.
```
+1 on the rendering of R^n -> R^m

dsaxton replied:
Actually, the "m" is a little hard to read (it looks a lot like the "n", so it may confuse people into thinking this is actually R^n -> R^n). What do you think? I think we could at least add some text like "and m is a positive integer" to bring attention to the fact that there are two different values here.

Also whichever we choose we probably want to do the same in the docstring for jackknife itself (currently it's rendered differently).

HDembinski replied:
Good points, maybe I should use `n` and `k`. Using unicode should be fine, it is supported everywhere nowadays and since we are not targeting Python2, I think it is safe to use here. I would try it out and roll back if people complain.

HDembinski replied:
Ah, I think I used m because superscript letters do not exist for all characters in unicode.

HDembinski commented:
It looks like the failed checks failed due to timeouts, can you restart them?

dsaxton commented:
> It looks like the failed checks failed due to timeouts, can you restart them?

Done
